### PR TITLE
Make EntryProcessor lambda-friendly

### PR DESCRIPTION
**Pull request by mmedenjak**
`AbstractEntryProcessor` and `EntryBackupProcessor` have been removed. Both partition owner and backup replica use `EntryProcessor`. The return value from the backup entry processor is ignored.

Now, instead of:
```java
        map.executeOnKey(key, new AbstractEntryProcessor<Integer, Employee>() {

            @Override
            public Object process(Map.Entry<Integer, Employee> entry) {
                Employee employee = entry.getValue();
                if (employee == null) {
                    employee = new Employee();
                }
                employee.setSalary(value);
                entry.setValue(employee);
                return null;
            }
        });
```
We can use lambda:
```java
        map.executeOnKey(key,
                entry -> {
                    Employee employee = entry.getValue();
                    if (employee == null) {
                        employee = new Employee();
                    }
                    employee.setSalary(value);
                    entry.setValue(employee);
                    return null;
                });
```

Also cleaned up some warnings and used JDK8 features.

Depends on: https://github.com/hazelcast/hazelcast/pull/14994
EE: https://github.com/hazelcast/hazelcast-enterprise/pull/2954
Fixes: https://github.com/hazelcast/hazelcast/issues/7028

### CODE REVIEW

File: hazelcast/src/main/java/com/hazelcast/core/IMap.java [hunk 1]
```diff
@@ -2840,7 +2844,8 @@ boolean tryLock(K key, long time, TimeUnit timeunit, long leaseTime, TimeUnit le
      * if the write-behind queue has reached its per-node maximum
      * capacity.
      */
-    Map<K, Object> executeOnEntries(EntryProcessor entryProcessor, Predicate predicate);
+    <R> Map<K, R> executeOnEntries(EntryProcessor<? super K, ? super V, R> entryProcessor,
+                                   Predicate<K, V> predicate);
```
Comment (cangencer) [inline hazelcast/src/main/java/com/hazelcast/core/IMap.java]: I think `? super K, ? super V` for `Predicate`?

Comment (mmedenjak) [reply hazelcast/src/main/java/com/hazelcast/core/IMap.java]: `Predicate` and `Projection` improvements are reserved for a separate PR to keep this one focused and to minimise the number of changed lines as it's already pretty big.

**Comment by mmedenjak** [comment]
run-lab-run

File: hazelcast/src/main/java/com/hazelcast/core/IMap.java [hunk 2]
```diff
@@ -2840,7 +2848,8 @@ boolean tryLock(K key, long time, TimeUnit timeunit, long leaseTime, TimeUnit le
      * if the write-behind queue has reached its per-node maximum
      * capacity.
      */
-    Map<K, Object> executeOnEntries(EntryProcessor entryProcessor, Predicate predicate);
+    <R> Map<K, R> executeOnEntries(@Nonnull EntryProcessor<? super K, ? super V, R> entryProcessor,
+                                   @Nonnull Predicate<K, V> predicate);
```
Comment (cangencer) [inline hazelcast/src/main/java/com/hazelcast/core/IMap.java]: `Predicate<? super K, ? super V>`

Comment (mmedenjak) [reply hazelcast/src/main/java/com/hazelcast/core/IMap.java]: `Predicate` and `Projection` improvements are reserved for a separate PR to keep this one focused and to minimise the number of changed lines as it's already pretty big.

File: hazelcast/src/main/java/com/hazelcast/map/EntryProcessor.java [hunk 3]
```diff
@@ -85,22 +95,35 @@
...
+     * that no backups should be made.
+     * <p>
+     * Note that there is a possibility which an {@code EntryProcessor} can see
+     * that a key exists but its backup processor may not find it at run time due
+     * to an unsent backup of a previous operation (e.g. a previous put). In
+     * those situations, Hazelcast internally/eventually will sync those owner
+     * and backup partitions so you will not lose any data.
+     * When coding an backup entry processor, you should take that case into
+     * account, otherwise {@link java.lang.NullPointerException}s can be seen
+     * since {@link java.util.Map.Entry#getValue()} may return null.
```
Comment (mustafaiman) [inline hazelcast/src/main/java/com/hazelcast/map/EntryProcessor.java]: Maybe add a line to say return value of backup entry processor is ignored. People may get confused as they do not know what should be returned from a backup entry processor.

**Comment by mmedenjak** [comment]
Merging, thanks for the reviews, guys. I have some minor finishing touches but I'll address those in a separate PR to avoid having more merge conflicts on this one and waiting for one more build.

**Pull request merged**